torch_geometric/loader/link_neighbor_loader.py

```diff
@@ -0,0 +1,307 @@
+from typing import (
+    Any,
+    Callable,
+    Iterator,
+    List,
+    Optional,
+    Sequence,
+    Tuple,
+    Union,
+)
+
+import torch
+from torch import Tensor
+from torch.utils.data import DataLoader
+
+from torch_geometric.data import Data, HeteroData
+from torch_geometric.loader.base import DataLoaderIterator
+from torch_geometric.loader.neighbor_loader import NeighborSampler
+from torch_geometric.loader.utils import filter_data, filter_hetero_data
+from torch_geometric.typing import (
+    EdgeType,
+    InputEdges,
+    NumNeighbors,
+    OptTensor,
+)
+from torch_geometric.utils import mask_to_index
+
+
+class LinkNeighborSampler(NeighborSampler):
+    def __init__(
+        self,
+        data: Union[Data, HeteroData],
+        input_edges: Tensor,
+        num_neighbors: NumNeighbors,
+        replace: bool = False,
+        directed: bool = True,
+        input_edge_type=Optional[EdgeType],
+        share_memory: bool = False,
+    ):
+        super().__init__(
+            data,
+            num_neighbors,
+            replace,
+            directed,
+            "link",  # prevents failing assert
+            share_memory)
+        self.input_edges = input_edges
+
+        if issubclass(self.data_cls, HeteroData):
+            assert_is_hetro_edge_type(input_edge_type)
+            self.input_edge_type = input_edge_type
+            self.start_node_type = input_edge_type[0]
+            self.end_node_type = input_edge_type[2]
+
+    def __call__(self, index: Union[List[int], Tensor]):
+
+        if issubclass(self.data_cls, Data):
+            # take start and end node from each edge then deduplicate
+            query_nodes = torch.cat(
+                [self.input_edges[0][index], self.input_edges[1][index]],
+                dim=0)
+            query_nodes, reverse_query_nodes = torch.unique(
+                query_nodes, return_inverse=True)
+
+            # get sampled graph
+            node, row, col, edge, _ = super().__call__(query_nodes)
+
+            # assume order of nodes in the new graph is same as input
+            assert all(query_nodes == node[:len(query_nodes)])
+            out_graph_nodes = torch.argsort(query_nodes)
```

Comment:
Any idea if there is a better way to do this? I do see https://discuss.pytorch.org/t/how-to-quickly-inverse-a-permutation-by-using-pytorch/116205 suggests a minor optimization, and there are a few on StackOverflow (I can benchmark some improvements in another MR), but I'm wondering if I'm missing a way to avoid this entirely.

Reply:
This looks good to me. I think it would be interesting to benchmark the speed here compared to not doing any `unique` call.

Furthermore, you may want to remove the `assert` statement since this will always be true by design. Using `all()` will also call a really slow Python iterator implementation.

Reply:
Okay fair enough - I added it incase something changes in the implementation of the base class, but our tests should also catch this.